internal/auth/authn/multiauth.go

```diff
@@ -0,0 +1,796 @@
+	OrganizationAssignmentTypePerUser = "perUser"
+)
+
+// contextKey is a custom type for context keys to avoid collisions
+type contextKey string
+
+// ParsedTokenCtxKey is the context key for the parsed JWT token
+const ParsedTokenCtxKey contextKey = "parsed_token"
+
+// GetParsedTokenFromContext retrieves the parsed JWT token from the context if it exists
+func GetParsedTokenFromContext(ctx context.Context) (jwt.Token, bool) {
+	if token, ok := ctx.Value(ParsedTokenCtxKey).(jwt.Token); ok {
+		return token, true
+	}
+	return nil, false
+}
+
+// TokenType represents the type of JWT authentication token
+type TokenType int
+
+const (
+	TokenTypeOIDC TokenType = iota
+	TokenTypeK8s
+)
+
+// AuthProviderService interface for auth provider operations
+type AuthProviderService interface {
+	ListAuthProviders(ctx context.Context, params api.ListAuthProvidersParams) (*api.AuthProviderList, api.Status)
+	GetAuthProviderByIssuerAndClientId(ctx context.Context, issuer string, clientId string) (*api.AuthProvider, api.Status)
+}
+
+// AuthProviderCacheKey is a composite key for caching auth providers
+type AuthProviderCacheKey struct {
+	Issuer   string
+	ClientId string
+}
+
+// MultiAuth implements authentication using multiple providers with issuer-based routing
+type MultiAuth struct {
+	// Static providers - initialized once at startup, mapped by issuer
+	staticProviders map[string]common.AuthNMiddleware // issuer -> provider mapping
+
+	// Service for dynamic auth providers
+	authProviderService AuthProviderService
+
+	// TLS config for OIDC provider connections
+	tlsConfig *tls.Config
+
+	// Logger for authentication operations
+	log logrus.FieldLogger
+
+	// Dynamic OIDC providers - issuer+clientId -> provider mapping
+	dynamicProviders   map[AuthProviderCacheKey]common.AuthNMiddleware
+	dynamicProvidersMu sync.RWMutex
+
+	// Control for background loader goroutine
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+// NewMultiAuth creates a new MultiAuth instance
+func NewMultiAuth(authProviderService AuthProviderService, tlsConfig *tls.Config, log logrus.FieldLogger) *MultiAuth {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	m := &MultiAuth{
+		staticProviders:     make(map[string]common.AuthNMiddleware),
+		authProviderService: authProviderService,
+		tlsConfig:           tlsConfig,
+		log:                 log,
+		dynamicProviders:    make(map[AuthProviderCacheKey]common.AuthNMiddleware),
+		ctx:                 ctx,
+		cancel:              cancel,
+	}
+
+	return m
+}
+
+// AddStaticProvider adds a static authentication provider with its issuer
+func (m *MultiAuth) AddStaticProvider(issuer string, provider common.AuthNMiddleware) {
+	m.staticProviders[issuer] = provider
+}
+
+// HasProviders returns true if any providers are configured
+func (m *MultiAuth) HasProviders() bool {
+	return len(m.staticProviders) > 0
+}
+
+// Start starts the background loader goroutine
+func (m *MultiAuth) Start() {
```

Comment:
if you would like to use ctx which make sense you can pass it here and create a child and store the childs cancelFn in m.